runtime/bf16_support.cu

```diff
@@ -16,6 +16,22 @@ __device__ __inline__ __bfloat __float2bfloat(const float);
 struct __align__(2) __bfloat {
   __bfloat() = default;
 
+  __bfloat(const volatile __bfloat& other) {
+    __x = other.__x;
+  }
+
+  __bfloat(const volatile __bfloat&& other) {
+    __x = other.__x;
+  }
+
+  __device__ void operator=(const volatile __bfloat& other) volatile {
```

Comment:
For my education, why "volatile" at the end of the line is necessary?

Reply:
I think it's because you'd hit a compile error assigning a volatile object to a non-volatile one unless the method was qualified.

My question is: will this automatically promote `other` to `volatile` in case you do `a = b` and neither are volatile? If so then any assignments of these types will be treated as volatile which seems not right. Would we need another method `__device__ void operator=(const __bfloat& other) {`? My understanding is that if you implemented one of the default operators, then all the default operators would not be created. But since there's no default volatile assignment operator does that apply? Also do you need the vol=nonvol and nonvol=vol cases like `operator=(const __bfloat& other) volatile {`?

Reply:
> My understanding is that if you implemented one of the default operators, then all the default operators would not be created.

good catch. I thought I can get away with volatile assignment operator but tried that and seems to be not the case.

so yes I'll need another set of overload for 

novol = vol
vol = novol
__bfloat(const __bfloat&)

Reply:
> I think it's because you'd hit a compile error assigning a volatile object to a non-volatile one unless the method was qualified.

Isn't it because assigning whatever value to a volatile object requires the `volatile` qualifier? At least, that's what I understood from this [doc](https://en.cppreference.com/w/cpp/language/member_functions#Member_functions_with_cv-qualifiers).

Reply:
Yes. `volatile` qualifier at the end marks that `self` pointed to instance is volatile. Hence you get volatile access to things like `(this->)__x`.